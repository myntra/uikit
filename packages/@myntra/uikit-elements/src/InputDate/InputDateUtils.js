import PropTypes from 'prop-types'
import { parse as baseParse, format as baseFormat } from 'date-fns'

export const RE_ISO_DATE = /^(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d+)|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d)|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d)$/

export const DateType = PropTypes.oneOfType([PropTypes.instanceOf(Date), PropTypes.string])

export const DateRangeType = PropTypes.oneOfType([
  PropTypes.shape({
    from: PropTypes.string,
    to: PropTypes.string
  }),
  PropTypes.shape({
    from: PropTypes.instanceOf(Date),
    to: PropTypes.instanceOf(Date)
  })
])

const padStartWithZeros = (value, digits = 2) => `${value}`.padStart(digits, '0')

/**
 * Create date instance.
 *
 * @export
 * @param {number} year
 * @param {number} month
 * @param {number} date
 * @returns {Date}
 */
export function UTCDate(year, month, date) {
  return new Date(
    `${padStartWithZeros(year, 4)}-${padStartWithZeros(month + 1)}-${padStartWithZeros(date)}T00:00:00.000Z`
  )
}

/**
 * Drop time component from Date.
 *
 * @export
 * @param {Date} date
 * @returns {Date}
 */
export function onlyDate(date) {
  if (typeof date.toDate === 'function') {
    date = date.toDate() // dayjs interop.
  }

  return UTCDate(date.getFullYear(), date.getMonth(), date.getDate())
}

export function parse(str, format) {
  if (str instanceof Date) return str
  if (typeof str !== 'string') return null
  if (!format) throw new Error(`'format' is required.`)
  if (RE_ISO_DATE.test(str)) return new Date(str)
  if (str.length !== format.length) throw new Error(`Date parse exception: unequal length`)

  const date = baseParse(str, format, new Date())

  if (date instanceof Date) {
    return onlyDate(date)
  }

  throw new Error(`Date parse exception: '${str}' does not match '${format}'`)
}

export function format(value, pattern) {
  if (!value) return value
  if (!pattern) return value
  if (value instanceof Date) return baseFormat(value, pattern)
  if (typeof value === 'string') return format(parse(value, pattern), pattern)

  const result = {}
  if (value.from) result.from = format(value.from, pattern)
  if (value.to) result.to = format(value.to, pattern)
  return result
}

export function isDateEqual(a, b) {
  if (!(a instanceof Date)) return false
  if (!(b instanceof Date)) return false
  if (a.getTime() === b.getTime()) return true

  return a.getFullYear() === b.getFullYear() && a.getMonth() === b.getMonth() && a.getDate() === b.getDate()
}
