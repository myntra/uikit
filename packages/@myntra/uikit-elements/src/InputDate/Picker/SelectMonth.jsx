import React from 'react'
import PropTypes from 'prop-types'

const MONTHS = 'Jan,Feb,Mar,Apr,May,Jun,Jul,Aug,Sep,Oct,Nov,Dec'.split(',')
function SelectMonth({ month, onMonthSelect, ...props }) {
  return (
    <div {...props}>
      <select value={month} onChange={event => onMonthSelect(Number(event.target.value))}>
        {MONTHS.map((item, index) => (
          <option value={index} key={item}>
            {item}
          </option>
        ))}
      </select>
    </div>
  )
}

SelectMonth.propTypes = {
  month: PropTypes.number.isRequired,
  onMonthSelect: PropTypes.func.isRequired
}

export default SelectMonth
