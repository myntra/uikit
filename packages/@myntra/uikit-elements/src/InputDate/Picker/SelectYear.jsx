import React from 'react'
import PropTypes from 'prop-types'

function SelectYear({ year, onYearSelect, ...props }) {
  let startYearValue = year - 10
  let endYearValue = year + 10
  const yearArray = []
  if (startYearValue < 0) {
    startYearValue = 0
    endYearValue = 20
  }
  for (let i = startYearValue; i < endYearValue; i++) {
    yearArray.push(i)
  }

  return (
    <div {...props}>
      <select value={year} onChange={event => onYearSelect(Number(event.target.value))}>
        {yearArray.map(item => (
          <option value={item} key={item}>
            {item}
          </option>
        ))}
      </select>
    </div>
  )
}

SelectYear.propTypes = {
  year: PropTypes.number.isRequired,
  onYearSelect: PropTypes.func.isRequired
}

export default SelectYear
