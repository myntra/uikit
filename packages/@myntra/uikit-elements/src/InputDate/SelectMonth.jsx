import React from 'react'
import PropTypes from 'prop-types'
import styles from './InputDate.css' //eslint-disable-line

const monthArr = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']
const MonthComponent = ({ month, onMonthSelect }) => (
  <div className={styles.selectWrapper}>
    <select value={month} onChange={event => onMonthSelect(Number(event.target.value))}>
      {monthArr.map((item, index) => (
        <option value={index} key={item}>
          {item}
        </option>
      ))}
    </select>
  </div>
)

MonthComponent.propTypes = {
  month: PropTypes.number.isRequired,
  onMonthSelect: PropTypes.func.isRequired
}

export default MonthComponent
