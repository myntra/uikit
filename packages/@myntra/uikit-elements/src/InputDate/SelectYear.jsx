import React from 'react'
import PropTypes from 'prop-types'
import styles from './InputDate.css' //eslint-disable-line

const YearComponent = ({ currentYear, onYearSelect }) => {
  let startYearValue = currentYear - 10
  let endYearValue = currentYear + 10
  const yearArray = []
  if (startYearValue < 0) {
    startYearValue = 0
    endYearValue = 20
  }
  for (let i = startYearValue; i < endYearValue; i++) {
    yearArray.push(i)
  }

  return (
    <div className={styles.selectWrapper}>
      <select value={currentYear} onChange={event => onYearSelect(Number(event.target.value))}>
        {yearArray.map(item => (
          <option value={item} key={item}>
            {item}
          </option>
        ))}
      </select>
    </div>
  )
}

YearComponent.propTypes = {
  currentYear: PropTypes.number.isRequired,
  onYearSelect: PropTypes.func.isRequired
}

export default YearComponent
