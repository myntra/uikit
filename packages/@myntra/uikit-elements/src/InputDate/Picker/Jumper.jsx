import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import { classnames } from '@myntra/uikit-utils'
import SelectYear from './SelectYear'
import SelectMonth from './SelectMonth'
import styles from './Jumper.module.css'
import { UTCDate } from '../InputDateUtils'
import { Icon } from '../../index.js'

/**
 Select year and month to jump to a date.

 @since 0.0.0
 @status REVIEWING
 @example
 <InputDatePicker.Jumper year={this.state.year || 2018} month={this.state.month || 0} offset={0} hasNext hasPrev
    onJump={date => this.setState({ year: date.getUTCFullYear(), month: date.getUTCMonth() })} />
 */
class Jumper extends PureComponent {
  static propTypes = {
    /** @private */
    className: PropTypes.string,
    year: PropTypes.number.isRequired,
    month: PropTypes.number.isRequired,
    offset: PropTypes.number.isRequired,
    /**
     * @function
     * @param {Date} date
     */
    onJump: PropTypes.func.isRequired,
    hasPrev: PropTypes.bool,
    hasNext: PropTypes.bool
  }

  handleJump = (year, month, diff = 0) => {
    const date = UTCDate(year, month, 1)

    date.setMonth(date.getUTCMonth() + diff - this.props.offset)

    this.props.onJump(date)
  }
  handleNext = () => this.handleJump(this.props.year, this.props.month, 1)
  handlePrev = () => this.handleJump(this.props.year, this.props.month, -1)
  handleYearSelect = year => this.handleJump(year, this.props.month)
  handleMonthSelect = month => this.handleJump(this.props.year, month)

  render() {
    const { year, month, offset, className, onJump, hasNext, hasPrev, ...props } = this.props

    return (
      <div {...props} className={classnames(className, 'jumper').use(styles)}>
        {hasPrev ? (
          <div className={classnames('prev').use(styles)} role="button" onClick={this.handlePrev}>
            <Icon name="chevron-left" />
          </div>
        ) : (
          <div className={classnames('prev', 'placeholder').use(styles)} />
        )}
        <SelectMonth
          month={month}
          onMonthSelect={this.handleMonthSelect}
          className={classnames('select').use(styles)}
        />
        <SelectYear className={classnames('select').use(styles)} year={year} onYearSelect={this.handleYearSelect} />
        {hasNext ? (
          <div className={classnames('next').use(styles)} role="button" onClick={this.handleNext}>
            <Icon name="chevron-right" />
          </div>
        ) : (
          <div className={classnames('next', 'placeholder').use(styles)} />
        )}
      </div>
    )
  }
}

Jumper.SelectMonth = SelectMonth
Jumper.SelectYear = SelectYear

export default Jumper
