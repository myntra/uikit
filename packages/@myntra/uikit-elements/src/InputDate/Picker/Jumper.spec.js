import React from 'react'
import { shallow } from 'enzyme'
import Jumper from './Jumper'
import SelectMonth from './SelectMonth'
import SelectYear from './SelectYear'
import { Icon } from '../..'
import { UTCDate } from '../InputDateUtils'

describe('render', () => {
  it('should render', () => {
    const wrapper = shallow(<Jumper year={2018} month={5} offset={0} onJump={jest.fn()} />)

    expect(wrapper.find(SelectMonth)).toBeTruthy()
    expect(wrapper.find(SelectYear)).toBeTruthy()
    expect(wrapper).toMatchSnapshot()
  })

  it('should render next/prev', () => {
    const wrapper = shallow(<Jumper year={2018} month={5} offset={0} onJump={jest.fn()} hasNext hasPrev />)
    const icons = wrapper.find(Icon)

    expect(icons).toHaveLength(2)
    expect(wrapper).toMatchSnapshot()
  })
})

describe('action', () => {
  it('year jump', () => {
    const fn = jest.fn()
    const wrapper = shallow(<Jumper year={2018} month={5} offset={0} onJump={fn} />)
    const year = wrapper.find(SelectYear)

    year.props().onYearSelect(2019)

    expect(fn).toHaveBeenCalledWith(UTCDate(2019, 5, 1))
  })

  it('month jump', () => {
    const fn = jest.fn()
    const wrapper = shallow(<Jumper year={2018} month={5} offset={0} onJump={fn} />)
    const month = wrapper.find(SelectMonth)

    month.props().onMonthSelect(2)

    expect(fn).toHaveBeenCalledWith(UTCDate(2018, 2, 1))
  })

  it('month next jump', () => {
    const fn = jest.fn()
    const wrapper = shallow(<Jumper year={2018} month={5} offset={0} onJump={fn} hasNext />)
    const month = wrapper.find('.next')

    month.simulate('click')

    expect(fn).toHaveBeenCalledWith(UTCDate(2018, 6, 1))
  })

  it('month prev jump', () => {
    const fn = jest.fn()
    const wrapper = shallow(<Jumper year={2018} month={5} offset={0} onJump={fn} hasPrev />)
    const month = wrapper.find('.prev')

    month.simulate('click')

    expect(fn).toHaveBeenCalledWith(UTCDate(2018, 4, 1))
  })
})
