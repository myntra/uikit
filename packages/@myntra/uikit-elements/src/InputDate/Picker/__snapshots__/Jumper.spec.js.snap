// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render should render 1`] = `
<div
  className="jumper"
>
  <div
    className="prev placeholder"
  />
  <SelectMonth
    className="select"
    month={5}
    onMonthSelect={[Function]}
  />
  <SelectYear
    className="select"
    onYearSelect={[Function]}
    year={2018}
  />
  <div
    className="next placeholder"
  />
</div>
`;

exports[`render should render next/prev 1`] = `
<div
  className="jumper"
>
  <div
    className="prev"
    onClick={[Function]}
    role="button"
  >
    <Icon
      name="chevron-left"
    />
  </div>
  <SelectMonth
    className="select"
    month={5}
    onMonthSelect={[Function]}
  />
  <SelectYear
    className="select"
    onYearSelect={[Function]}
    year={2018}
  />
  <div
    className="next"
    onClick={[Function]}
    role="button"
  >
    <Icon
      name="chevron-right"
    />
  </div>
</div>
`;
